jxl/src/api.rs

```diff
@@ -0,0 +1,351 @@
+    /// Parses an ICC profile, returning a ColorEncoding and whether the ICC profile represents a
+    /// CMYK profile.
+    fn parse_icc(&mut self, icc: &[u8]) -> Result<(ColorEncoding, bool)>;
+
+    /// Initializes `n` transforms (different transforms might be used in parallel) to
+    /// convert from color space `input` to colorspace `output`, assuming an intensity of 1.0 for
+    /// non-absolute luminance colorspaces of `intensity_target`.
+    /// It is an error to not return `n` transforms.
+    fn initialize_transforms(
+        &mut self,
+        n: usize,
+        max_pixels_per_transform: usize,
+        input: JxlColorProfile,
+        output: JxlColorProfile,
+        intensity_target: f32,
+    ) -> Result<Vec<Box<dyn JxlCmsTransformer>>>;
+}
+
+pub enum JxlProgressiveMode {
+    /// Renders all pixels in every call to Process.
+    Eager,
+    /// Renders pixels once passes are completed.
+    Pass,
+    /// Renders pixels only once the final frame is ready.
+    FullFrame,
+}
+
+#[non_exhaustive]
+pub struct JxlDecoderOptions {
+    pub adjust_orientation: bool,
+    pub unpremultiply_alpha: bool,
+    pub render_spot_colors: bool,
+    pub coalescing: bool,
+    pub desired_intensity_target: Option<f32>,
+    pub skip_preview: bool,
+    pub progressive_mode: JxlProgressiveMode,
+}
+
+impl Default for JxlDecoderOptions {
+    fn default() -> Self {
+        Self {
+            adjust_orientation: true,
+            unpremultiply_alpha: false,
+            render_spot_colors: true,
+            coalescing: true,
+            skip_preview: false,
+            desired_intensity_target: None,
+            progressive_mode: JxlProgressiveMode::Pass,
+        }
+    }
+}
+
+mod states {
+    pub trait JxlState {}
+    pub struct Initialized;
+    pub struct WithImageInfo;
+    pub struct WithFrameInfo;
+    impl JxlState for Initialized {}
+    impl JxlState for WithImageInfo {}
+    impl JxlState for WithFrameInfo {}
+}
+
+use states::*;
+
+#[allow(dead_code)]
+struct JxlDecoderInner {
+    // TODO more fields
+    options: JxlDecoderOptions,
+    cms: Option<Box<dyn JxlCms>>,
+}
+
+// Q: do we plan to add support for box decoding?
+// If we do, one way is to take a callback &[u8; 4] -> Box<dyn Write>.
+
+pub struct JxlDecoder<State: JxlState> {
+    inner: JxlDecoderInner,
+    _state: PhantomData<State>,
+}
+
+impl<S: JxlState> JxlDecoder<S> {
+    fn wrap_inner(inner: JxlDecoderInner) -> Self {
+        Self {
+            inner,
+            _state: PhantomData,
+        }
+    }
+
+    /// Resets entirely a decoder, producing a new decoder with the same settings.
+    /// This is faster than creating a new decoder in some cases.
+    pub fn reset(self) -> JxlDecoder<Initialized> {
+        JxlDecoder::wrap_inner(self.inner)
+    }
+
+    /// Rewinds a decoder to the start of the file, allowing past frames to be displayed again.
+    pub fn rewind(self) -> JxlDecoder<Initialized> {
+        JxlDecoder::wrap_inner(self.inner)
```

Comment:
Is `rewind` also producing a new decoder? If so perhaps perhaps we only need one of those?

Reply:
Rewind will optimize decoding assuming that _the same image_ will be decoded.

Reply:
ok, perhaps worth mentioning in the doc string, but not super important